### ISSUE TITLE
`bugprone-infinite-loop` incorrectly claims infinite loop

### ISSUE THREAD

**Issue by bbannier** [llvm/llvm-project#88111]
For `do`/`while` loops with unchanged loop condition the loop (while pointless) will execute exactly once, but `clang-tidy`'s `bugprone-infinite-loop` check incorrectly claims that the loop would be infinite.

```cpp
int main() {
    bool changed = false;  // Diagnostic disappears if declared `constexpr`.
    do {  // This loop is infinite; none of its condition variables (changed) are updated in the loop body
    } while ( changed );
}
```

**Comment by HerrCai0907** [comment]
Theoretically it is possible to support this feature in clang-tidy. But it needs to introduce cfg / value analysis which will cause performance issue.

**Comment by bbannier** [comment]
> Theoretically it is possible to support this feature in clang-tidy. But it needs to introduce cfg / value analysis which will cause performance issue.

If this check cannot do control flow analysis for performance reasons it should be tweaked to avoid false positives. That being said, I would argue that for such a check to be useful that it needs to do some form of flow analysis, e.g., the following produces no warning (false negative):

```cpp
bool changed = false;
do { 
    changed = true;
} while ( changed ); // no diagnostic.
```

or even worse:

```cpp
do {
} while ( true ); // no diagnostic.
```
